imblearn/base.py

```diff
@@ -63,7 +63,7 @@ def sample(self, X, y):
 
         return self._sample(X, y)
 
-    def fit_sample(self, X, y):
+    def fit_sample(self, X, y, **fit_params):
```

Comment:
I think that I would avoid to pass a keyword here.
We need to pass the keyword when creating the instance.

In addition, I would propose to have an argument `categorical_features`='auto'` or `categorical_features=[1, ...]` to inform for the name of the index. With `'auto'`, think that we can check the column which are binary {0, 1} previously encoded by a `OneHotEncoder`.
It will be slower and we can directly pass the index of the column then (or even the column name if we want to support pandas dataframe).

Reply:
Hi there, thanks for looking into this PR!

1. **fit_params keyword** (or, more specifically, *feature_indices* keyword parameter in SMOTENC class): I also considered the option of making it a __init__ parameter to be set when SMOTENC instance is being created. However, in the end I opted for a keyword argument. Here's my motivation: *feature_indices* indicates which columns of input represent the nominal features after they were one-hot-encoded. Thus, it is directly connected to the provided input *X*. So if you want to call *fit* method on the different *X*, you might have to provide different *feature_indices*. Speaking more generally, I assumed that __init__ should accept parameters which are **not** input-dependent and *fit* can take parameters which are dependent on given input *X*.

2. **categorical_features=auto**: I like this idea! (I guess you mean aforementioned *feature_indices* parameter?). Just one question: would it maybe make sense to implement it as an improvement/follow-up ticket after this PR is merged? Asking since I'm not sure if I will have enough time in near future to work on it.

As for you question regarding OneHotEncoding you're right, it is not mentioned in the paper. Also, strictly speaking, it is not part of SMOTENC implementation either. It's just expects the data to be one-hot encoded because: a) SMOTENC re-uses a lot of code from SMOTE (nearest-neighbors, initial sample generation etc.) which will not work if input contains non-numerical data ; b) OneHotEncoding is a go-to tool when dealing with nominal features in sklearn; c) when applied, it conveniently returns *feature_indices* array to be used by SMOTENC

Regards,
Denis

Reply:
1. My concern is that we don't follow the current scikit-learn API. The `ColumnTransformer` will for instance get the same kind of parameters at initialization. Even if I agree that it would make sense to have it at fit time, I am not aware of mechanism to be able to forward the information of the `OneHotEncoder` to `SMOTE` within a `Pipeline`. It is something that at the long term should be done with `sample_props` which is actually something pass in `fit` similarly to what you are proposing. However, in the meantime I would be more conservative.

2. I prefer the term `categorical_features` instead of `features_indices` since it is explicit what are we targeting. My proposal regarding the `'auto'` is actually to partially solve the problem spotted in 1. Regarding the extra work, I could implement it and push it in your PR, if you are OK with it.

3. I agree with your last point. However, since that we are actually not implemented the original `SMOTE-NC`, I would propose to rename the class to something different to not have any issue with other implementation which might differ. I don't have a good naming yet `CatSMOTE` maybe. Any proposal would be nice.

Reply:
1. Agreed (I'll address it during re-base)
2. Agreed
3. I still think that we could use SMOTE-NC as a name. Even though OneHotEncoding is not mentioned in the paper, I think using it does not contradict the paper. All the steps 1-3 (median computation, nearest neighbors, sample generation) behave as described in the paper. I would call the usage of OneHotEncoder just an "implementation detail". But maybe I'm missing something so please let me know if you have a different idea.
(for reference https://arxiv.org/pdf/1106.1813.pdf, section 6.1)

Reply:
@here I just wanted to ask if a method has been implemented to inverse_transform the OneHotEncoded values after applying the SMOTE-NC.

Also, I did test it with some sample datasets having categorical values and nominal, but it results in a way that the one-hot-encoded values (binary) are also getting incremented with the 'step' parameter provided.It results in the values such as (0.1,0.12,0.6..etc). 

How to convert them back to the original categorical dataset with these type of values?

Reply:
>  I just wanted to ask if a method has been implemented to inverse_transform the OneHotEncoded values after applying the SMOTE-NC.

Since I integrated the OneHotEncode within SMOTENC, the feature will be reverted.

> How to convert them back to the original categorical dataset with these type of values?

Now you will get the original categories so it should not be an issue. I still have to implement more tests. to be sure that it behave as exptected in all situation.

Reply:
@glemaitre There is some issue with OHE (when using normally). 
Recent version provides functionality of converting it without LabelEncoding it. I tried the process with a dataset of shape ((284807, 30)) and it was giving me memory error.

I hope it might be of some use.